File: go/q-21-merge-two-sorted-lists/q-21-merge-two-sorted-lists.go
package main

import "fmt"

//21.合并两个有序链表.merge-two-sorted-lists

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}
	if list2 == nil {
		return list1
	}

	var ans *ListNode
	if list1.Val < list2.Val {
		ans = list1
		list1 = list1.Next
	} else {
		ans = list2
		list2 = list2.Next
	}

	cur := ans
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			cur.Next = list1
			list1 = list1.Next
		} else {
			cur.Next = list2
			list2 = list2.Next
		}
		cur = cur.Next
	}

	if list1 != nil {
		cur.Next = list1
	}
	if list2 != nil {
		cur.Next = list2
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	var l1 *ListNode = nil
	var l2 *ListNode = nil
	var result = mergeTwoLists(l1, l2)
	fmt.Println(result)
}

File: cpp/q-21-merge-two-sorted-lists/q-21-merge-two-sorted-lists.cpp
#include "iostream"
using namespace std;

//21.合并两个有序链表.merge-two-sorted-lists

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (!list1) return list2;
        if (!list2) return list1;
        ListNode *t, *ans;
        if (list1->val < list2->val) {
            t = list1;
            list1 = list1->next;
        } else {
            t = list2;
            list2 = list2->next;
        }
        ans = t;
        while (list1 && list2) {
            if (list1->val < list2->val) {
                t->next = list1;
                list1 = list1->next;
            } else {
                t->next = list2;
                list2 = list2->next;
            }
            t = t->next;
        }
        if (list1) t->next = list1;
        if (list2) t->next = list2;
        return ans;
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(){
    ListNode *l1 = nullptr, *l2 = nullptr;
    cout << (new Solution())->mergeTwoLists(l1, l2) << endl;
    return 0;
}
